scripts/esm/grid.mjs

```diff
@@ -0,0 +1,319 @@
+        gridUV.y = invertLine.y ? gridUV.y : 1.0 - gridUV.y;
+        vec2 grid2 = smoothstep(drawWidth + lineAA, drawWidth - lineAA, gridUV);
+
+        grid2 *= clamp(targetWidth / drawWidth, 0.0, 1.0);
+        grid2 = mix(grid2, targetWidth, clamp(uvDeriv * 2.0 - 1.0, 0.0, 1.0));
+        grid2.x = invertLine.x ? 1.0 - grid2.x : grid2.x;
+        grid2.y = invertLine.y ? 1.0 - grid2.y : grid2.y;
+
+        return mix(grid2.x, 1.0, grid2.y);
+    }
+
+    void main(void) {
+        vec2 uv = uv0;
+
+        vec2 pos = (uv * 2.0 - 1.0) * uHalfExtents;
+        vec2 ddx = dFdx(pos);
+        vec2 ddy = dFdy(pos);
+
+        float epsilon = 1.0 / 255.0;
+
+        vec2 levelPos;
+        float levelSize;
+        float levelAlpha;
+
+        levelPos = pos * 0.1;
+        levelSize = 2.0 / 1000.0;
+        levelAlpha = pristineGrid(levelPos, ddx * 0.1, ddy * 0.1, vec2(levelSize));
+        if (levelAlpha > epsilon) {
+            vec3 color;
+            if (abs(levelPos.x) < levelSize) {
+                if (abs(levelPos.y) < levelSize) {
+                    color = vec3(1.0);
+                } else {
+                    color = uColorZ;
+                }
+            } else if (abs(levelPos.y) < levelSize) {
+                color = uColorX;
+            } else {
+                color = vec3(0.9);
+            }
+            gl_FragColor = vec4(color, levelAlpha);
+            return;
+        }
+
+        levelPos = pos;
+        levelSize = 1.0 / 100.0;
+        levelAlpha = pristineGrid(levelPos, ddx, ddy, vec2(levelSize));
+        if (levelAlpha > epsilon) {
+            if (uResolution < 1) {
+                discard;
+            }
+            gl_FragColor = vec4(vec3(0.7), levelAlpha);
+            return;
+        }
+
+        levelPos = pos * 10.0;
+        levelSize = 1.0 / 100.0;
+        levelAlpha = pristineGrid(levelPos, ddx * 10.0, ddy * 10.0, vec2(levelSize));
+        if (levelAlpha > epsilon) {
+            if (uResolution < 2) {
+                discard;
+            }
+            gl_FragColor = vec4(vec3(0.7), levelAlpha);
+            return;
+        }
+
+        discard;
+    }
+`;
+
+class Grid extends Script {
+    /**
+     * @type {number}
+     */
+    static RESOLUTION_LOW = 0;
+
+    /**
+     * @type {number}
+     */
+    static RESOLUTION_MEDIUM = 1;
+
+    /**
+     * @type {number}
+     */
+    static RESOLUTION_HIGH = 2;
+
+    /**
+     * @type {ShaderMaterial}
+     * @private
+     */
+    _material;
+
+    /**
+     * @type {MeshInstance}
+     * @private
+     */
+    _meshInstance;
+
+    /**
+     * @type {Vec2}
+     * @private
+     */
+    _halfExtents = new Vec2();
+
+    /**
+     * @type {Color}
+     * @private
+     */
+    _colorX = new Color(1, 0.3, 0.3);
+
+    /**
+     * @type {Color}
+     * @private
+     */
+    _colorZ = new Color(0.3, 0.3, 1);
+
+    /**
+     * @type {number}
+     * @private
+     */
+    _resolution = Grid.RESOLUTION_HIGH;
+
+    constructor(args) {
```

Comment:
what are the args this constructor accepts? Typically, a script is not constructed as such right?

Reply:
@marklundin - this is a bit strange, is this how users are expected to handle this in the scripts?

Reply:
There's nothing inherently wrong with this, but `initialize()` is the more canonical use here. Is there a specific reason why this shouldn't happen in the initialize phase @kpal81xd?

Reply:
It's the `script.create(Grid, { properties })` which adds properties in the `initialize` phase. Linking back to the original discussion here https://github.com/playcanvas/engine/issues/6785. Let's discuss